### PR TITLE
Optimize transformations

### PR DESCRIPTION
**Pull request by nbirillo**

### CODE REVIEW

**Comment by nbirillo** [comment]
Probably, we need to implement something like [Cache replacement policies](https://en.wikipedia.org/wiki/Cache_replacement_policies) to clear the cache partially if it is too big, what do you think, @eartser?

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/distance/measurers/gumtree/GumTreeDistanceMeasurer.kt [hunk 1]
```diff
@@ -43,8 +47,11 @@ class GumTreeDistanceMeasurerByPsi(
...
     private val psiManager: PsiManager = PsiManager.getInstance(project),
 ) : GumTreeDistanceMeasurerBase() {
-    override fun String.parseTree(): TreeContext =
-        this.asPsiFile(Language.PYTHON, psiManager) {
-            getTreeContext(it)
+    private val codeToTreeContext: HashMap<String, TreeContext> = HashMap()
+    override fun String.parseTree(): TreeContext {
+        val trimmedCode = this.trimCode()
+        return codeToTreeContext[trimmedCode] ?: run {
+            this.asPsiFile(Language.PYTHON, psiManager) { getTreeContext(it) }
```
Comment (eartser) [inline code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/distance/measurers/gumtree/GumTreeDistanceMeasurer.kt:54]: Should we put new tree context into the hashmap here?

**Review by eartser — CHANGES_REQUESTED** [review]
Thanks! One small comment:

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/distance/measurers/gumtree/GumTreeDistanceMeasurer.kt (hunk 1 above)
Comment (nbirillo) [reply code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/distance/measurers/gumtree/GumTreeDistanceMeasurer.kt:54]: oh yes, sorry

**Pull request merged**